src/main/java/com/example/pokebattlez/battle/controller/Battle.java

```diff
@@ -17,56 +20,66 @@
 
     private BattleService battleService;
-
-    public void setBattleService(BattleService battleService) {
-        this.battleService = battleService;
-    }
+    private static final BattleUtils battleUtils = new BattleUtils(); // Ask about this
 
     public void registerPokemonTeam(Long trainerId, List<Pokemon> pokemon) {
         this.pokemon.put(trainerId, pokemon);
     }
 
-    public void switchPokemonTo(Long trainerId, int position) {
+    public void registerSwitchPokemon(Long trainerId, int position) {
         playerActions.add(PlayerAction.builder()
                 .trainerId(trainerId)
                 .switchToPosition(position)
-                .action(PlayerAction::switchPokemon)
+                .action(Battle::switchPokemon)
                 .priority(1000)
                 .build()
         );
     }
 
-    public void useMoveInPosition(Long trainerId, int position) {
+    public void registerUseMove(Long trainerId, int position) {
         playerActions.add(PlayerAction.builder()
                 .trainerId(trainerId)
                 .moveInPosition(position)
-                .action(PlayerAction::useMove)
+                .action(Battle::useMove)
                 .priority(activePokemon.get(trainerId).getStats().get(Stat.SPEED).getValue())
                 .build()
         );
     }
 
     public void cancelAction(Long trainerId) {
-        playerActions.stream().filter(playerAction -> playerAction.getTrainerId().equals(trainerId)).findFirst().ifPresent(playerActions::remove);
+        playerActions.stream()
+                .filter(playerAction -> playerAction.trainerId.equals(trainerId)).findFirst()
+                .ifPresent(playerActions::remove);
     }
 
-    @Data
-    @Builder
-    private static class PlayerAction {
-        private int priority;
+    private static void switchPokemon(Battle battle, PlayerAction playerAction) {
+        battle.pokemon.get(playerAction.trainerId).stream()
+                .filter(pokemon1 -> pokemon1.getPosition() == playerAction.switchToPosition).findFirst()
+                .ifPresent(pokemon1 -> battle.activePokemon.replace(playerAction.trainerId, pokemon1));
+        battle.playerActions.remove(playerAction);
+    }
 
-        private Long trainerId;
-        private int switchToPosition;
-        private int moveInPosition;
+    private static void useMove(Battle battle, PlayerAction playerAction) {
+        // TODO
+    }
+
+    private int calculateDamage(Pokemon attacker, Pokemon defender, Pokemon.Move move) {
+        return (int) Math.floor((Math.floor(
+                Math.floor(
+                        (Math.floorDiv(Math.floorDiv((Math.floorDiv(attacker.getLevel() * 2, 5) + 2) * battleUtils.getAttackTypeValueForMoveCategory(attacker, move.getCategory()) * move.getPower(), battleUtils.getDefenceTypeValueForMoveCategory(defender, move.getCategory()) ), 50) + 2) * battleUtils.stabModifier(move.getType(), attacker.getTypes())
```

Comment:
Would it be possible to dissect this formula into some logical parts? Just asking...

Reply:
((2A/5+2)*B*C)/D)/50)+2)*X)*Y/10)*Z)/255

A = attacker's Level
B = attacker's Attack or Special
C = attack Power
D = defender's Defense or Special
X = same-Type attack bonus (1 or 1.5)
Y = Type modifiers (40, 20, 10, 5, 2.5, or 0)
Z = a random number between 217 and 255

This is the formula, personally, I didn't see how to separate it in any logical way...